tests/integration_tests/test_public_api.py

```diff
@@ -0,0 +1,145 @@
+import ipaddress
+from contextlib import contextmanager, suppress
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Generator
+
+import pytest
+
+import pycloudlib
+from pycloudlib.cloud import BaseCloud
+from pycloudlib.instance import BaseInstance
+from pycloudlib.util import LTS_RELEASES
+
+cloud_config = """\
+#cloud-config
+runcmd:
+  - echo 'hello' >> /var/tmp/example.txt
+"""
+
+
+@pytest.fixture
+def cloud(request):
+    cloud_instance: BaseCloud = request.param(
+        tag="pycl-test",
+        timestamp_suffix=True,
+    )
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.upload_key(
+            public_key_path=cloud_instance.config["public_key_path"],
+            private_key_path=cloud_instance.config["private_key_path"],
+            name=cloud_instance.tag,
+        )
+
+    yield cloud_instance
+
+    if isinstance(cloud_instance, pycloudlib.EC2):
+        cloud_instance.delete_key(name=cloud_instance.tag)
+
+
+@contextmanager
+def launch_instance(
+    cloud_instance: BaseCloud, **kwargs
+) -> Generator[BaseInstance, None, None]:
+    instance = cloud_instance.launch(**kwargs)
+    try:
+        yield instance
+    finally:
+        instance.delete()
+
+
+def assert_example_output(instance: BaseInstance):
+    example_output = instance.execute("cat /var/tmp/example.txt").stdout
+    assert example_output == "hello"
+
+
+def exercise_push_pull(instance: BaseInstance):
+    with TemporaryDirectory() as tmpdir:
+        push_path = Path(tmpdir).joinpath("pushed")
+        push_path.write_text("pushed", encoding="utf-8")
+        instance.push_file(str(push_path), "/var/tmp/pushed")
+        assert (
+            "pushed" == instance.execute("cat /var/tmp/pushed").stdout.strip()
+        )
+
+        instance.execute("echo 'pulled' > /var/tmp/pulled")
+        pull_path = Path(tmpdir).joinpath("pulled")
+        instance.pull_file("/var/tmp/pulled", str(pull_path))
+        assert pull_path.read_text(encoding="utf-8").strip() == "pulled"
+
+
+def exercise_instance(instance: BaseInstance):
+    assert instance.name is not None
+    assert ipaddress.ip_address(instance.ip)
+
+    with suppress(NotImplementedError):
+        ip_address = instance.add_network_interface()
+        try:
+            assert ipaddress.ip_address(ip_address)
+        finally:
+            instance.remove_network_interface(ip_address)
+
+    exercise_push_pull(instance)
+
+    assert_example_output(instance)
+    boot_id = instance.get_boot_id()
+    instance.restart()
+    assert boot_id != instance.get_boot_id()
+
+    assert_example_output(instance)
+    instance.shutdown()
+
+    instance.start()
+    assert_example_output(instance)
+
+
+@pytest.mark.parametrize(
+    "cloud",
+    [
+        pytest.param(
+            pycloudlib.Azure, id="azure", marks=pytest.mark.main_check
```

Comment:
What's preventing you from doing that? You can run `pytest "tests/integration_tests/test_public_api.py::test_public_api[lxd_container]"` or `tox -e integration-tests -- "tests/integration_tests/test_public_api.py::test_public_api[lxd_container]"` if you want to run the lxd_container-specific verison.

I'm not really expecting all that many "general purpose" tests. Beyond this, I would think any cloud-specific tests would be in their own directory.

Reply:
that works for me, thanks for calling out the specific option there for platform-specialized set of unit tests. I was thinking toward having jenkins runners that would be opinionated based on platform support like we have for cloud-init so that if we would know specifically from the cloud platform "flavor" of CI/CD job  we'd know whether we have a problem that only happens on one platform... But we can just as easily click into a failed job log or failure artifacts to understand the breadth of the failure. No need to over-engineer the solution we have currently. thanks